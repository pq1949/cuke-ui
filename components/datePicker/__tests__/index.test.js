import moment from "moment";
import React from "react";
import assert from "power-assert";
import { render, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import DatePicker from "../index";
import Button from "../../button";

describe("<DatePicker/>", () => {
  it("should render a <DatePicker/> components", () => {
    const wrapper = render(<DatePicker />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("should render find .cuke-date-picker classnames", () => {
    const wrapper = shallow(<DatePicker />);
    assert(wrapper.find(".cuke-date-picker").length === 1);
  });
  it("should render find .cuke-date-picker-loading classnames", () => {
    const wrapper = shallow(<DatePicker loading />);
    assert(wrapper.find(".cuke-date-picker-loading").length === 1);
  });
  it("should render today button", () => {
    const wrapper = shallow(<DatePicker />);
    assert(wrapper.find(".cuke-date-picker-footer-today").length === 1);
  });
  it("should cannot render today button", () => {
    const wrapper = shallow(<DatePicker showToday={false} />);
    assert(wrapper.find(".cuke-date-picker-footer-today").length === 0);
  });
  it("should cannot trigger click event when disabled", () => {
    const onChange = jest.fn();
    const wrapper = shallow(<DatePicker disabled onChange={onChange} />);
    wrapper.find(".cuke-date-picker").simulate("click");
    expect(onChange).not.toHaveBeenCalled();
  });
  it("should render extraFooter", () => {
    const Footer = () => <Button>1</Button>;
    const wrapper = shallow(<DatePicker extraFooter={<Footer />} />);
    assert(wrapper.find(Footer).length === 1);
    assert(wrapper.find(".cuke-date-picker-footer-extra").length === 1);
  });
  it("should can not render today and clear when showToday and showClear is false", () => {
    const wrapper = shallow(<DatePicker showToday={false} showClear={false} />);
    assert(wrapper.find(".cuke-date-picker-footer-today").length === 0);
    assert(wrapper.find(".cuke-date-picker-footer-clear").length === 0);
  });
  it.skip("should trigger panel visible change", () => {
    const onPanelVisibleChange = jest.fn();
    const wrapper = shallow(
      <DatePicker onPanelVisibleChange={onPanelVisibleChange} />
    );
    wrapper.find(".cuke-date-picker-input").simulate("click");
    expect(onPanelVisibleChange).toHaveBeenCalled();
  });

  it("should cannot trigger panel visible change when disabled", () => {
    const onPanelVisibleChange = jest.fn();
    const wrapper = shallow(
      <DatePicker disabled onPanelVisibleChange={onPanelVisibleChange} />
    );
    wrapper.find(".cuke-date-picker-input").simulate("click");
    expect(onPanelVisibleChange).not.toHaveBeenCalled();
  });

  it("should trigger today date", () => {
    const wrapper = shallow(<DatePicker />);
    wrapper.find(".cuke-date-picker-footer-today").simulate("click");
    expect(
      wrapper
        .state()
        .momentSelected.day(0)
        .minute(0)
        .second(0)
        .millisecond(0)
    ).toEqual(
      moment()
        .day(0)
        .minute(0)
        .second(0)
        .millisecond(0)
    );
  });

  it("should render prev month when subtract btn clicked", () => {
    const wrapper = shallow(<DatePicker />);
    wrapper
      .find(".cuke-date-picker-switch-group")
      .at(0)
      .simulate("click");
    expect(
      wrapper
        .state()
        .momentSelected.day(0)
        .minute(0)
        .second(0)
        .millisecond(0)
    ).toEqual(
      moment()
        .subtract(1, "month")
        .day(0)
        .minute(0)
        .second(0)
        .millisecond(0)
    );
  });

  it("should render add month when subtract btn clicked", () => {
    const wrapper = shallow(<DatePicker />);
    wrapper
      .find(".cuke-date-picker-switch-group")
      .at(1)
      .simulate("click");
    expect(
      wrapper
        .state()
        .momentSelected.day(0)
        .minute(0)
        .second(0)
        .millisecond(0)
    ).toEqual(
      moment()
        .add(1, "month")
        .day(0)
        .minute(0)
        .second(0)
        .millisecond(0)
    );
  });

  it("should clear date when clear btn clicked", () => {
    const wrapper = shallow(<DatePicker />);
    wrapper.find(".cuke-date-picker-footer-clear").simulate("click");
    expect(
      wrapper
        .state()
        .momentSelected.day(0)
        .minute(0)
        .second(0)
        .millisecond(0)
    ).toEqual(
      moment()
        .day(0)
        .minute(0)
        .second(0)
        .millisecond(0)
    );
  });

  it("should window cannot trigger click event when un mount", () => {
    const onClick = jest.fn();
    const wrapper = shallow(<DatePicker />);
    wrapper.unmount();
    window.onclick = () => onClick;
    expect(onClick).not.toHaveBeenCalled();
  });
  it("should set new moment selected value when did update ", () => {
    const wrapper = shallow(<DatePicker value={moment()} />);
    wrapper.setProps({
      value: moment("1996-08-25")
    });
    wrapper.update();
    expect(wrapper.state().momentSelected).toEqual(moment("1996-08-25"));
  });
  it("should can not update when set equal value ", () => {
    const wrapper = shallow(<DatePicker value={moment("1996-08-25")} />);
    wrapper.setProps({
      value: moment("1996-08-25")
    });
    wrapper.update();
    expect(wrapper.state().isSelectedMoment).toEqual(true);
  });
  it("should can not render any text when not show prev and next month day ", () => {
    const wrapper = shallow(
      <DatePicker
        value={moment("2018-11-25")}
        showDayInPrevMonth={false}
        showDayInNextMonth={false}
      />
    );
    expect(
      wrapper
        .find(".cuke-date-picker-last-month")
        .at(0)
        .text()
    ).toContain("");
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("should from 11 month => 10 month when last month button clicked", () => {
    const wrapper = shallow(
      <DatePicker
        value={moment("2018-11-25")}
      />
    );
    wrapper
    .find(".cuke-date-picker-last-month")
    .at(0)
    .simulate('click')
    expect(
      wrapper.state().momentSelected.month()
    ).toEqual(10)
  });

  it("should can set default Value & defaultValue.date is equal state.selectedDate", () => {
    const wrapper = shallow(<DatePicker value={moment("2018-11-25")} />);
    expect(wrapper.state().selectedDate).toEqual(moment("2018-11-25").date());
  });

  it("should render custom bottom position", () => {
    const wrapper = shallow(<DatePicker position="bottom" />);
    assert(wrapper.find(".cuke-date-picker-position-bottom").length === 1);
  });

  it("should render custom top position", () => {
    const wrapper = shallow(<DatePicker position="top" />);
    assert(wrapper.find(".cuke-date-picker-position-top").length === 1);
  });
});