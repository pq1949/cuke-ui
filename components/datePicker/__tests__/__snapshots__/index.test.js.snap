// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker/> should can not render any text when not show prev and next month day  1`] = `
<div
  className="cuke-date-picker cuke-date-picker-position-bottom"
  onChange={[Function]}
  value={"2018-11-24T16:00:00.000Z"}
>
  <div
    className="cuke-date-picker-inner"
  >
    <Input
      className="cuke-date-picker-input"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      placeholder="请选择"
      prefixCls="cuke-input"
      readonly={true}
      type="text"
      value="2018-11-25"
    />
    <DownIcon
      className="cuke-date-picker-arrow"
    />
  </div>
  <div
    className="cuke-date-picker-content cuke-date-picker-close cuke-ui-no-animate"
    style={
      Object {
        "left": undefined,
        "top": undefined,
      }
    }
  >
    <Spin
      indicator={<LoadingIcon />}
      prefixCls="cuke-spin"
      size="large"
      spinning={false}
      tip=""
    >
      <div
        className="cuke-date-picker-header"
      >
        <span
          className="cuke-date-picker-date"
        >
          2018
          年 
            
          11
          月
        </span>
        <span
          className="cuke-date-picker-switch"
        >
          <span
            className="cuke-date-picker-switch-group"
            onClick={[Function]}
          >
            <ArrowLeftIcon />
          </span>
          <span
            className="cuke-date-picker-switch-group"
            onClick={[Function]}
          >
            <ArrowRightIcon />
          </span>
        </span>
      </div>
      <div
        className="cuke-date-picker-items"
      >
        <div>
          <span
            className="cuke-date-picker-item cuke-date-picker-day-title"
            key="Su"
          >
            Su
          </span>
          <span
            className="cuke-date-picker-item cuke-date-picker-day-title"
            key="Mo"
          >
            Mo
          </span>
          <span
            className="cuke-date-picker-item cuke-date-picker-day-title"
            key="Tu"
          >
            Tu
          </span>
          <span
            className="cuke-date-picker-item cuke-date-picker-day-title"
            key="We"
          >
            We
          </span>
          <span
            className="cuke-date-picker-item cuke-date-picker-day-title"
            key="Th"
          >
            Th
          </span>
          <span
            className="cuke-date-picker-item cuke-date-picker-day-title"
            key="Fr"
          >
            Fr
          </span>
          <span
            className="cuke-date-picker-item cuke-date-picker-day-title"
            key="Sa"
          >
            Sa
          </span>
        </div>
        <span
          className="cuke-date-picker-item cuke-date-picker-last-month"
          key="first-date-0"
          onClick={[Function]}
        />
        <span
          className="cuke-date-picker-item cuke-date-picker-last-month"
          key="first-date-1"
          onClick={[Function]}
        />
        <span
          className="cuke-date-picker-item cuke-date-picker-last-month"
          key="first-date-2"
          onClick={[Function]}
        />
        <span
          className="cuke-date-picker-item cuke-date-picker-last-month"
          key="first-date-3"
          onClick={[Function]}
        />
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-0"
          onClick={[Function]}
        >
          1
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-1"
          onClick={[Function]}
        >
          2
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-2"
          onClick={[Function]}
        >
          3
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-3"
          onClick={[Function]}
        >
          4
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-4"
          onClick={[Function]}
        >
          5
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-5"
          onClick={[Function]}
        >
          6
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-6"
          onClick={[Function]}
        >
          7
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-7"
          onClick={[Function]}
        >
          8
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-8"
          onClick={[Function]}
        >
          9
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-9"
          onClick={[Function]}
        >
          10
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-10"
          onClick={[Function]}
        >
          11
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-11"
          onClick={[Function]}
        >
          12
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-12"
          onClick={[Function]}
        >
          13
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-13"
          onClick={[Function]}
        >
          14
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-14"
          onClick={[Function]}
        >
          15
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-15"
          onClick={[Function]}
        >
          16
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-16"
          onClick={[Function]}
        >
          17
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-17"
          onClick={[Function]}
        >
          18
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-18"
          onClick={[Function]}
        >
          19
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-19"
          onClick={[Function]}
        >
          20
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-20"
          onClick={[Function]}
        >
          21
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-21"
          onClick={[Function]}
        >
          22
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-22"
          onClick={[Function]}
        >
          23
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-23"
          onClick={[Function]}
        >
          24
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month cuke-date-picker-selected-date"
          key="current-date-24"
          onClick={[Function]}
        >
          25
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-25"
          onClick={[Function]}
        >
          26
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-26"
          onClick={[Function]}
        >
          27
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-27"
          onClick={[Function]}
        >
          28
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-28"
          onClick={[Function]}
        >
          29
        </span>
        <span
          className="cuke-date-picker-item cuke-date-picker-current-month"
          key="current-date-29"
          onClick={[Function]}
        >
          30
        </span>
      </div>
      <div
        className="cuke-date-picker-footer cuke-date-picker-has-border"
      >
        <div
          className="cuke-date-picker-footer-today"
          onClick={[Function]}
        >
          今天
        </div>
        <div
          className="cuke-date-picker-footer-clear"
          onClick={[Function]}
        >
          清除
        </div>
      </div>
    </Spin>
  </div>
</div>
`;

exports[`<DatePicker/> should render a <DatePicker/> components 1`] = `
<div
  class="cuke-date-picker cuke-date-picker-position-bottom"
>
  <div
    class="cuke-date-picker-inner"
  >
    <input
      class="cuke-input cuke-date-picker-input"
      placeholder="请选择"
      readonly=""
      type="text"
      value=""
    />
    <svg
      class="cuke-date-picker-arrow"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
  <div
    class="cuke-date-picker-content cuke-date-picker-close cuke-ui-no-animate"
  >
    <div
      class="cuke-date-picker-header"
    >
      <span
        class="cuke-date-picker-date"
      >
        2019年   1月
      </span>
      <span
        class="cuke-date-picker-switch"
      >
        <span
          class="cuke-date-picker-switch-group"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="15 18 9 12 15 6"
            />
          </svg>
        </span>
        <span
          class="cuke-date-picker-switch-group"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="cuke-date-picker-items"
    >
      <div>
        <span
          class="cuke-date-picker-item cuke-date-picker-day-title"
        >
          Su
        </span>
        <span
          class="cuke-date-picker-item cuke-date-picker-day-title"
        >
          Mo
        </span>
        <span
          class="cuke-date-picker-item cuke-date-picker-day-title"
        >
          Tu
        </span>
        <span
          class="cuke-date-picker-item cuke-date-picker-day-title"
        >
          We
        </span>
        <span
          class="cuke-date-picker-item cuke-date-picker-day-title"
        >
          Th
        </span>
        <span
          class="cuke-date-picker-item cuke-date-picker-day-title"
        >
          Fr
        </span>
        <span
          class="cuke-date-picker-item cuke-date-picker-day-title"
        >
          Sa
        </span>
      </div>
      <span
        class="cuke-date-picker-item cuke-date-picker-last-month"
      >
        30
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-last-month"
      >
        31
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        1
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month cuke-date-picker-selected-date"
      >
        2
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        3
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        4
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        5
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        6
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        7
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        8
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        9
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        10
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        11
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        12
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        13
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        14
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        15
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        16
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        17
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        18
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        19
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        20
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        21
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        22
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        23
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        24
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        25
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        26
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        27
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        28
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        29
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        30
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-current-month"
      >
        31
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-next-month"
      >
         1
      </span>
      <span
        class="cuke-date-picker-item cuke-date-picker-next-month"
      >
         2
      </span>
    </div>
    <div
      class="cuke-date-picker-footer cuke-date-picker-has-border"
    >
      <div
        class="cuke-date-picker-footer-today"
      >
        今天
      </div>
      <div
        class="cuke-date-picker-footer-clear"
      >
        清除
      </div>
    </div>
  </div>
</div>
`;
